app/src/util/itertools.ts

```diff
@@ -84,6 +100,46 @@ function* filter<T>(predicate: (x: T) => boolean, input: Generator<T>): Generato
   }
 }
 
+function* dropwhile<T>(predicate: (x: T) => boolean, input: Generator<T>): Generator<T> {
+  for (const item of input) {
+    if (!predicate(item)) {
+      yield item;
+      break;
+    }
+  }
+  yield* input;
+}
+
+function* dropuntil<T>(predicate: (x: T) => boolean, input: Generator<T>): Generator<T> {
+  for (const item of input) {
+    if (predicate(item)) {
+      yield item;
+      break;
+    }
+  }
+  yield* input;
+}
+
+function* takewhile<T>(predicate: (x: T) => boolean, input: Generator<T>): Generator<T> {
+  for (const item of input) {
+    if (predicate(item)) {
+      yield item;
+    } else {
+      break;
+    }
+  }
+}
+
+function* takeuntil<T>(predicate: (x: T) => boolean, input: Generator<T>): Generator<T> {
```

Comment:
mmh i dont really understand what these functions ought to do

Reply:
takeuntil is similar to pythons [itertools.takeuntil](https://docs.python.org/3/library/itertools.html#itertools.takewhile): it yields elements from the start of generator until the predicate is matched for the first time and then drops the rest of the iterator.

Reply:
I think the usage is quite useful as an example of how these functions work and how they play together:
https://github.com/audapolis/audapolis/pull/145/files#diff-7969c210d8aa3bf9b6e9c5ebfd871e8f63f436ce9972f3f32cd3c793018f4fb5R346

We have our `generator`, which contains all the Items of the document, sorted by they absoluteStart. We want to collect all items meet a certain condition (in this case: `(x.absoluteStart - EPSILON) < time < (x.absoluteStart + x.length + EPSILON)`, i.e. the play time is within this element). Since we know that all elements that meet this condition appear in a row within out generator, we only want to get the first block of matching items. Image the following, where each `F` means one items that does not match and each `T` means one item that does match:

```
FFFFFFTTFFFFFFFF
```

We only want to get the two `T` items, so we first remove all the `F`s before them using `generator.dropuntil(CONDITION)` and are left with:

```
TTFFFFFFFF
``` 

now we use `generator.takewhile(CONDITION)` to only get the first block of items that matches the `CONDITION` and we are left with

```
TT
```

which is all the items we need.

In this usecase, this saves us from fully draining the `generator` and checking every items, as we do until now, because the `takewhile` function drops the rest of the iterator once its condition is no longer met. The `*until` and `*while`-functions work in opposite ways: `takeuntil` yields elements until the condition is met for the first time and then drops the rest (i.e. stops the iteration). `takewhile` yields elements until the condition is not met for the first time and then drops the rest

Reply:
thanks for the great explanation :)